src/hardware/dac.rs

```diff
@@ -69,6 +69,40 @@ use hal::dma::{
 static mut DAC_BUF: [[[u16; SAMPLE_BUFFER_SIZE]; 3]; 2] =
     [[[0; SAMPLE_BUFFER_SIZE]; 3]; 2];
 
+/// Custom type for referencing DAC output codes.
+/// The internal integer is the raw code written to the DAC output register.
+#[derive(Copy, Clone)]
+pub struct DacCode(pub u16);
+
+impl Into<f32> for DacCode {
+    fn into(self) -> f32 {
+        // The output voltage is generated by the DAC with an output range of +/- 4.096 V. This
+        // signal then passes through a 2.5x gain stage. Note that the DAC operates using unsigned
+        // integers, and u16::MAX / 2 is considered zero voltage output. Thus, the dynamic range of
+        // the output stage is +/- 10.24 V. At a DAC code of zero, there is an output of -10.24 V,
+        // and at a max DAC code, there is an output of (slightly less than) 10.24 V.
+
+        let dac_volts_per_lsb = 10.24 * 2.0 / u16::MAX as f32;
```

Comment:
The scaling is wrong again ;) Here and the same for `DacCode`.
And `u16::MAX/2` is not zero voltage output. Code `(1 << 15)` is.

It's much better to let the compiler do the math and just write down the electronics.


```suggestion
        let dac_volts_per_lsb = 4.096 * 2.5 / (1u16 << 15) as f32;
```